base_layer/core/src/proof_of_work/monero_rx.rs

```diff
@@ -103,16 +172,69 @@ fn monero_difficulty_calculation(header: &BlockHeader) -> Result<Difficulty, Mer
     }
 }
 
-fn create_input_blob(_data: &MoneroData) -> Result<String, MergeMineError> {
-    // Todo deserialize monero data to create string for  randomX vm
-    // Returning an error here so that difficulty can return 0 as this is not yet implemented.
-    Err(MergeMineError::HashingError)
+fn create_input_blob(data: &MoneroData) -> Result<String, MergeMineError> {
+    #[cfg(feature = "monero_merge_mining")]
+    {
+        let header = serialize::<MoneroBlockHeader>(&data.header.clone());
+        let mut count = serialize::<VarInt>(&VarInt(1 + data.count as u64));
+        let mut hashes = Vec::new();
+        for item in &data.transaction_hashes {
+            hashes.push(Hash(from_slice(item.clone().as_bytes())));
+        }
+        let mut root = tree_hash(hashes);
+        let mut encode2 = header;
+        encode2.append(&mut root);
+        encode2.append(&mut count);
+        Ok(hex::encode(encode2).into())
+    }
+    #[cfg(not(feature = "monero_merge_mining"))]
+    {
+        Err(MergeMineError::HashingError)
+    }
+}
+
+// Utility function to transform array to fixed array
+fn from_slice(bytes: &[u8]) -> [u8; 32] {
+    let mut array = [0; 32];
+    let bytes = &bytes[..array.len()]; // panics if not enough data
+    array.copy_from_slice(bytes);
+    array
+}
+
+fn verify_root(monero_data: &MoneroData) -> Result<(), MergeMineError> {
+    let mut hashes = Vec::new();
+    for item in &monero_data.transaction_hashes {
+        hashes.push(Hash(from_slice(item.to_vec().as_slice())));
+    }
+    let root = tree_hash(hashes);
+
+    if !(monero_data.transaction_root.to_vec() == root) {
+        return Err(MergeMineError::ValidationError);
+    }
+    Ok(())
+}
+
+pub fn merged_mining_subfield(header: &BlockHeader) -> SubField {
+    let hash = header.merged_mining_hash();
+    let mut tag = [0; 32];
+    tag.copy_from_slice(&hash[..32]);
+    let depth = 0;
+    SubField::MergeMining(VarInt(depth), Hash::hash(&tag))
 }
 
-fn verify_header(_header: &BlockHeader, _monero_data: &MoneroData) -> Result<(), MergeMineError> {
-    // todo
-    // verify that our header is in coinbase
-    // todo
-    // verify that coinbase is in root.
+fn verify_header(header: &BlockHeader, monero_data: &MoneroData) -> Result<(), MergeMineError> {
```

Comment:
Perhaps we need to add a validate option to the header field and call that from the validators. 

Currently the blake headers dont have much to validate, so all the validation happens in the get_acheved_difficulty. 

With the monero merge mined headers, there is actual verification required. I think its perhaps going to be cleaner that way and if we need to add future validation for standard headers, it can happen there then.